fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModJarMetadata.java

```diff
@@ -9,34 +9,49 @@
 import net.minecraftforge.forgespi.locating.IModFile;
 import java.lang.module.ModuleDescriptor;
 import java.util.Objects;
-
 import org.jetbrains.annotations.ApiStatus;
 
 @ApiStatus.Internal
 public final class ModJarMetadata implements JarMetadata {
+    private static final String AUTOMATIC_MODULE_NAME = "Automatic-Module-Name";
     private IModFile modFile;
+    private String name;
+    private String version;
     private ModuleDescriptor descriptor;
 
-    ModJarMetadata() {
-    }
+    ModJarMetadata() { }
 
     public void setModFile(IModFile file) {
         this.modFile = file;
+        var mods = this.modFile.getModFileInfo().getMods();
+
+        if (!mods.isEmpty()) {
+            var main = mods.get(0);
+            this.name = main.getModId();
+            this.version = main.getVersion().toString();
+        }
+
+        var jar = file.getSecureJar();
+        var aname = jar.moduleDataProvider().getManifest().getMainAttributes().getValue(AUTOMATIC_MODULE_NAME);
+        if (aname != null)
+            this.name = aname;
     }
 
     @Override
     public String name() {
-        return modFile.getModFileInfo().moduleName();
+        return this.name;
     }
```

Comment:
I actually ran into this issue because FMLModContainer passes in the ModFile and not the module, but then gets the module to make the mod instance.
So with the changes I made to ModFileInfo, it now calls this function. So unless a custom language loader breaks this loading this name() method should be the authority. Tested it on a couple mod packs with kotlin mods and it works fine.